transit_odp/pipelines/pipelines/dataset_etl/transform.py

```diff
@@ -201,3 +309,97 @@ def get_most_common_districts(stops: pd.DataFrame):
     @staticmethod
     def get_most_common_localities(stops: pd.DataFrame):
         return get_most_common_localities(stops)
+
+    def sync_flexible_zone(self, flexible_stop_points):
+        filtered_stop_points = flexible_stop_points[
+            (flexible_stop_points["bus_stop_type"] == "flexible")
+            & (~flexible_stop_points["naptan_id"].isna())
+        ]
+        filtered_stop_points_naptan_id = filtered_stop_points["naptan_id"].tolist()
+        flexiblezone_qs = FlexibleZone.objects.filter(
+            naptan_stoppoint_id__in=filtered_stop_points_naptan_id
+        )
+        flexible_zone_df = create_naptan_flexible_zone_df_from_queryset(flexiblezone_qs)
+        if not flexible_zone_df.empty:
+            filtered_stop_points_naptan_id = filtered_stop_points.reset_index().merge(
+                flexible_zone_df.reset_index(),
+                how="left",
+                left_on="naptan_id",
+                right_on="naptan_id",
+            )
+            filtered_stop_points_naptan_id = filtered_stop_points_naptan_id[
+                ["atco_code", "flexible_location"]
+            ]
+            filtered_stop_points_naptan_id.set_index("atco_code", inplace=True)
+            return filtered_stop_points_naptan_id
+        return pd.DataFrame()
+
+    def create_flexible_timing_link(self, flexible_journey_details, flexible_jp_to_jps):
+        def get_stop_sequence(group, jp_section_ref):
+            group["order"] = range(len(group))
+            group["route_link_ref"] = (
+                str(jp_section_ref) + "RL" + (group["order"] + 1).astype(str)
+            )
+            group["jp_timing_link_id"] = (
+                str(jp_section_ref) + "TL" + (group["order"] + 1).astype(str)
+            )
+            group["from_stop_ref"] = group["atco_code"].shift(0)
+            group["to_stop_ref"] = group["atco_code"].shift(-1)
+            return group.dropna(subset=["to_stop_ref"])
+
+        if not flexible_journey_details.empty and not flexible_jp_to_jps.empty:
+            flexible_jp_to_jps = flexible_jp_to_jps.reset_index()[
+                ["journey_pattern_id", "jp_section_ref"]
+            ]
+            journey_patterns = flexible_journey_details.reset_index().merge(
+                flexible_jp_to_jps,
+                how="left",
+                left_on="journey_pattern_id",
+                right_on="journey_pattern_id",
+            )
+            grouped = journey_patterns.groupby("jp_section_ref")
```

Comment:
We could pass the function `get_stop_sequence` as an apply function after `groupby` to return expected data to avoid `for` iteration below. `transform_service_pattern_stop` can be used to for reference